### PR TITLE
ci(triage): match test as a name segment, not just a whole component

### PR DESCRIPTION
**Pull request by beardthelion**
#202 merged on 2026-08-10 and fixed the namespaced-attribute case, so main already detects `#[tokio::test]` and `#[sqlx::test]`. This branch was opened against the old base for the same bug and has been rebased down to the part that is still missing.

What is left is the harness whose final path component carries `test` as an underscore segment. `#[test_case(1)]` and `#[wasm_bindgen_test]` fall through main's `test\b` and wrongly earn the PR a `needs-tests` label. The final component becomes:

```
(?:[A-Za-z0-9]+_)*test(?:_[A-Za-z0-9]+)*
```

keeping the leading `::`, the `r#` form, and the whitespace tolerance #202 added.

Two things worth stating outright, since neither is obvious from the diff.

**It is a segment, not a substring.** An earlier version of this branch used `\w*test\w*`. That also matches `#[contest]` and `#[latest]`, and because `addsInlineTest` feeding `touchedTests` means a match SUPPRESSES the label, a false positive here silently clears `needs-tests` on a PR that added no tests. The heuristic has to fail in the loud direction. The price is the harness names with no underscore, `#[rstest]` among them, which stay unmatched on purpose and produce the corrigible failure instead.

**The character classes are `[A-Za-z0-9]` rather than `\w` for a specific reason.** `\w` contains `_`, which makes `(?:_\w+)*` an ambiguous repetition that backtracks exponentially. On this workflow that is reachable: it triggers on `pull_request_target` with `pull-requests: write`, so the patch text comes from whoever opened the PR and is evaluated on a privileged runner. With `\w`, `#[test` followed by a long `_a` run measured 1.07ms at 16 characters, 5.59ms at 18, 16.09ms at 20 and 71.08ms at 22, so a line under 90 characters stalls the job to its timeout. The literal-delimiter form measures 0.007ms on the same input and stays linear across every adversarial shape tried.

Scope note: this changes no verdict on real history. Over 972 file-patches from the last 400 commits touching `crates/**/*.rs`, the old and new regexes agree everywhere, and none of `rstest`, `test_case`, `wasm_bindgen_test`, `::tokio::test` or `r#test` appears in the tree. It is a net for harnesses we do not use yet, not a fix for a live mislabel.

CodeRabbit's note about multiline comments and raw strings defeating a patch-line regex is accurate but unchanged here, and applies equally to what is on main today. Tracking it separately rather than growing this into a lexer, since the label is advisory.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved detection of Rust inline tests during pull request triage.
  * More reliably recognizes test attributes across whitespace, line breaks, comments, qualified paths, raw identifiers, and parameterized forms.
  * Avoids false positives from unrelated context, removed lines, delimiters, and excessively long continuations.
  * Continues to limit detection to Rust files and relevant added patch lines.

* **Tests**
  * Added automated validation covering standard, edge-case, and malformed patch content.

### CODE REVIEW

**Review by jatmn — CHANGES_REQUESTED** [review]
I found one item that needs maintainer input before this merges.

## Findings

- [P2] Resolve overlap with open PR #202 before merging
  `.github/workflows/pr-triage.yml` (same file, competing fix)
  Open PR [#202](https://github.com/Gitlawb/node/pull/202) (`fix/pr-triage-async-test-detection`, +208/−4) fixes the same `needs-tests` inline-test detector, closes #201, and already went through multiple review rounds (lexical stripping of comments/strings/raw literals, head-file lexing across hunks, blob API fetch, bounded-scan inconclusive handling, `#[r#test]` / `#[::tokio::test]`). This PR (+15/−1) is a regex-only subset of that work. Only one should land; merging both would conflict. Please close or supersede #202 explicitly, or close this PR in favor of #202 if the fuller approach is preferred.

**Review by jatmn — CHANGES_REQUESTED** [review]
I rechecked the current head. The regex change does what it needs to for this repo: it now catches added `#[tokio::test]` and `#[sqlx::test]` lines that the old pattern missed (the cases that falsely triggered `needs-tests` on #275 and #262), while still matching bare `#[test]` and `#[cfg(test)]`. I do not see actionable code defects in the diff itself.

One item still needs maintainer input before this merges.

## Findings

- [P2] Resolve overlap with open PR #202 before merging
  `.github/workflows/pr-triage.yml` (same file, competing fix)
  Open PR [#202](https://github.com/Gitlawb/node/pull/202) (`fix/pr-triage-async-test-detection`, +7/−1 on current head) still fixes the same `needs-tests` inline-test detector, closes #201, and has already been through multiple review rounds. This PR is a different regex-only take on the same line. Only one should land; merging both would conflict. Please close or supersede #202 explicitly, or close this PR in favor of #202 if that approach is preferred. Note that #202 was simplified since my earlier review (it is no longer the larger lexer-based change), but the conflict is unchanged. The two regexes make different tradeoffs (#277 matches broader harness names like `#[rstest]` / `#[test_case(1)]` via substring matching; #202 matches `#[::tokio::test]` and `#[r#test]` forms that this regex misses), so the choice is not a pure duplicate even though the scope is.

## Notes (not findings)

- CodeRabbit's inline comment about multiline comments and raw strings is a fair description of a patch-line-regex limitation, but both competing PRs accept that tradeoff for an advisory label. I do not treat it as a merge blocker.
- Substring `test` matching (e.g. `#[contest]`) is an intentional breadth choice to catch future harness macros without enumerating them; none of those decoy attributes appear in the tree, and invalid attributes would still fail `cargo` checks.

**Comment by beardthelion** [comment]
#202 merged on 2026-08-10, so the overlap this was blocked on is resolved by that landing rather than by a choice between the two. I have rebased this branch onto current main and cut it down to the piece #202 does not cover.

Main's `test\b` requires the final path component to be exactly `test`. A harness that carries `test` as an underscore segment still falls through, so `#[test_case(1)]` and `#[wasm_bindgen_test]` get a `needs-tests` label they should not. The diff now widens only that, keeping the `::` prefix, `r#`, and whitespace handling you reviewed on #202.

Two notes on the current head, since both changed the design.

Your point about substring breadth was right and I dropped it. The earlier `\w*test\w*` also matched `#[contest]` and `#[latest]`, and since a match suppresses the label rather than adding it, a false positive is the silent direction: a testless PR clears its own label. It is now an underscore-delimited segment, which costs `#[rstest]` and keeps the failure loud.

The classes are `[A-Za-z0-9]` rather than `\w` because `\w` contains `_`, which makes the repetition ambiguous and backtracks exponentially. This workflow runs on `pull_request_target`, so the patch text is fork-supplied on a privileged runner. Measured with `\w`: `#[test` plus a long `_a` run took 1.07ms at 16 chars, 16.09ms at 20, 71.08ms at 22, roughly 4x per two characters, so a line under 90 characters runs the job to its timeout. The shipped form measures 0.007ms there and stays linear on every shape I threw at it. That was mine, introduced during this rebase and fixed before pushing; nothing on main has it.

Worth being blunt about the value: across 972 file-patches from the last 400 commits, old and new agree on every one, and none of these harness names appears in the tree. This is a net for later, not a fix for a live mislabel. If you would rather not carry the extra regex surface for that, closing it is a reasonable call and I will not argue it.

On CodeRabbit's multiline-comment and raw-string note, I agree with your read. It is real, it applies equally to main today, and it needs a lexer rather than a wider regex, so I would keep it out of this PR.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P3] Preserve legal test attributes with comments after the path
  `.github/workflows/pr-triage.yml:96`
  The new `[ \t]*[\]\(]` suffix turns the detector's formerly token-boundary-based match into one that requires the next token to be `]` or `(` after horizontal whitespace only. Rust permits comments and line breaks in that position: `#[test /* rationale */]` compiles and runs, but its added patch line now returns false (the base `test\b` expression returns true). A Rust-only PR using that valid spelling is therefore labeled `needs-tests` and receives the guidance comment even though it added a test.

  Address the root cause by making the post-path check recognize Rust token separation rather than just horizontal whitespace plus two delimiters. Keep the detector anchored to added Rust lines and retain the intentional bounded, non-ambiguous matching behavior; add direct regression cases for a comment and newline after bare and namespaced test paths so a future false-positive hardening change does not again reject legal attributes.

**Comment by beardthelion** [comment]
Rechecked at 11053204. The suffix now treats block comments, line breaks, and split-line closers as Rust token separation before ] or (, so `#[test /* … */]`, namespaced forms, and a path split across added patch lines no longer false-trigger `needs-tests`. Live-regex matrix: 32/32. ReDoS guard: 64 probes, still under budget.

Left `#[rstest]` unmatched on purpose (a false positive here suppresses the label). Multiline-comment/raw-string limitation unchanged, still advisory-only.

**Review by jatmn — CHANGES_REQUESTED** [review]
I rechecked the current head and found issues that need to be addressed before this is ready.

## Findings

- [P3] Bind the split-line closer to its attribute path
  `.github/workflows/pr-triage.yml:106`
  The two `.test(p)` calls lose the relationship between the path-only attribute and its continuation: the first can match `+#[test_case` anywhere in `f.patch`, while the second can match any added line beginning `]` or `(` anywhere else, regardless of order, distance, or hunk. This is reproducible with a complete, compiling, testless Rust change—not only malformed source: put `#[test_case` on a line inside an added raw-string fixture and add an unrelated tuple expression whose `(` starts a later added line. Current `main` returns `addsInlineTest=false` for that patch, but this head returns true through the fallback, so `touchedTests` silently suppresses/removes `needs-tests`. That is the failure direction the comments on lines 83-87 explicitly say must remain loud, and it confirms CodeRabbit's current line-107 request.

  Please address the lost ordering/association rather than adding another independent patch-wide predicate. Parse the patch as ordered diff records or use one bounded match that consumes the specific, immediately following diff-prefixed continuation for the path it matched. Preserve the direct inline branch and legitimate `+#[test_case` / `+(1)]` splits, but add regressions where delimiter-looking added lines occur before the path, later in the same hunk, and in another hunk. This does not require solving the workflow's accepted raw-string/comment lexical limitation; the required outcome is that unrelated patch records cannot complete each other.

- [P3] Preserve legal Rust comment separators after the test path
  `.github/workflows/pr-triage.yml:100`
  Replacing main's `test\b` boundary with a mandatory `TEST_ATTR_SEP` plus `]`/`(` regresses legal tests because this is only a partial Rust separator grammar. It has no `//` branch, and `/\*[^*]*\*+(?:[^/*][^*]*\*+)*/` necessarily stops at the first `*/` even though Rust block comments nest. Both `#[test // rationale\n]` and `#[test /* outer /* inner */ outer */]` compile and register under `rustc --test`; the exact merge-base/live-target detector returns true for both, while this head returns false. A Rust-only PR using either spelling is therefore newly labeled `needs-tests` and receives the persistent guidance comment despite adding a real test. The green PR checks do not cover this behavior: there is no committed matrix for the inline `github-script`, and `pull_request_target` runs the target branch's workflow rather than exercising this proposed workflow body.

  Please fix the root compatibility issue rather than continuing to enumerate comment spellings in a flat regex. One bounded option is to retain main's established branch for the already-supported exact terminal `test` forms and apply the stricter suffix/terminator logic only where the new underscore-segment names need it; another is a small token-aware separator scan that handles `//`-to-newline and nested block-comment depth. Either approach should keep the new `test_case` / `wasm_bindgen_test` behavior, the intentional `rstest` exclusion, and the linear-time/fail-loud constraints. Add focused cases for ordinary and nested block comments, line comments, same-line and split closers, plus the existing adversarial runtime probes so this repair does not reopen the false-positive or ReDoS paths.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Keep non-code `test_case` text from clearing the test signal
  `.github/workflows/pr-triage.yml:155`
  `patchAddsInlineTest` starts a candidate scan at every added patch line, but it has no Rust lexical state that says whether that line is source code, a raw string, or a multiline comment. For example, this is valid, testless Rust:

  ```rust
  const FIXTURE: &str = r#"
  #[test_case
  (1)]
  "#;
  ```

  Its patch has immediately adjoining `+#[test_case` and `+(1)]` records, so the new scanner returns true. That makes `touchedTests` true and omits/removes `needs-tests`, even though no test was added. The merge-base `test\b` detector returns false for exactly this patch because it did not recognize underscore-suffixed harness names; this is therefore newly reachable through this PR, rather than the older general limitation of a patch-line heuristic. Address the root cause by making newly supported attribute forms fail loud when they occur in lexical non-code, and add regressions for raw strings and multiline comments. Preserve the current bounded, linear-time scan and its strict association with adjoining added patch records; an implementation approach is intentionally left open.

- [P3] Correct the detector’s failure-direction contract
  `scripts/test-pr-triage-detect.mjs:16`
  The new matrix says that a false negative silently suppresses `needs-tests`, but the workflow does the opposite: `patchAddsInlineTest === false` leaves `touchedTests` false, so line 181 applies the label. A false positive is the silent failure because it sets `touchedTests` and clears the label. The test behavior itself is correct—uncertain input must return false—but the rationale contradicts both that behavior and the workflow’s stated fail-loud invariant. Correct the root safety contract in the comment, explicitly distinguishing false positives from false negatives, while retaining the existing uncertain-input expectation and cases.

**Comment by beardthelion** [comment]
Both findings from the latest review are addressed at d800f08.

**[P2] Keep non-code `test_case` text from clearing the test signal**

Added a lexical-state pre-pass to `patchAddsInlineTest` that tracks raw strings (`r#"`, `br#"`, `cr#"`), nested block comments, regular strings (which can span newlines), and line comments across consecutive added patch lines. An added line that starts inside any of these is skipped as a candidate, so `#[test_case]` appearing in a raw string or block comment no longer sets `touchedTests`.

The reviewer's exact fixture (`r#"\n#[test_case\n(1)]\n"#`) now returns false. Seven regression cases cover the reject path (raw string, block comment, no-hash raw string, multiline regular string) and the accept path (each construct closes, then a real test attribute on the next added line).

State is only tracked across consecutive added lines and reset at any break (context line, removal, hunk header). Context lines carry a partial view of the file, so a `"` or `r#"` or `/*` there might be inside a string that opened before the hunk's visible window. Scanning context lines would mark every later added line as non-code and miss real `#[test]` attributes (a false negative, reproduced on real commit bf4d8cd9). The tradeoff is that a non-code region opened by a context line is invisible: an added `#[test]` inside an existing raw string or block comment is still scanned. That is a false positive in the silent direction, but the old detector had the same behavior, the pattern is rare, and `needs-tests` is advisory.

**[P3] Correct the detector's failure-direction contract**

The comment in `scripts/test-pr-triage-detect.mjs` now states that false positives silently suppress `needs-tests` (they set `touchedTests` and clear the label), while false negatives apply the label loudly. The existing uncertain-input expectation (return false) is retained.

**Verification**

- `node scripts/test-pr-triage-detect.mjs` -> 38 cases pass
- 998 real patches across 500 commits compared against the old detector: 0 regressions, 0 new detections
- YAML validated
- 44-case edge probe covering byte strings, C strings, nested block comments, two-hunk state reset, and `attrCloserFollows` interaction with non-code in continuation lines
- Revert-checks confirm each guard is load-bearing

File: .github/workflows/pr-triage.yml [hunk 1]
```diff
@@ -76,9 +76,24 @@ jobs:
...
+            //
+            // The segment classes are [A-Za-z0-9]+ rather than \w+ on purpose, so that `_`
+            // is only ever the literal delimiter. \w+ contains `_`, which makes the repeat
+            // ambiguous and backtracks exponentially: on this workflow's pull_request_target
+            // trigger f.patch is fork-controlled, so `#[test` followed by a long `_a` run is
+            // a permissionless stall of the triage job.
             const addsInlineTest = files.some(f =>
               f.filename.endsWith(".rs") && f.patch &&
-              /^\+[ \t]*#\[[ \t]*(?:cfg[ \t]*\([ \t]*test[ \t]*\)|(?:::[ \t]*)?(?:[\w-]+[ \t]*::[ \t]*)*(?:r#)?test\b)/m.test(f.patch));
+              /^\+[ \t]*#\[[ \t]*(?:cfg[ \t]*\([ \t]*test[ \t]*\)|(?:::[ \t]*)?(?:[\w-]+[ \t]*::[ \t]*)*(?:r#)?(?:[A-Za-z0-9]+_)*test(?:_[A-Za-z0-9]+)*[ \t]*[\]\(])/m.test(f.patch));
```
[the remark this answers is not in the document]
Comment (beardthelion) [reply .github/workflows/pr-triage.yml]: Declined by design. `#[rstest]` has no underscore-delimited `test` segment, so matching it requires broadening the pattern to a substring match. That would also match `#[contest]` and `#[latest]`, which are not test attributes. A false positive here silently suppresses `needs-tests`, so the conservative exclusion is intentional. The test matrix has a `rstest stays excluded` case expecting false.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found one correctness issue worth fixing.

## Findings

- [P3] Recognize character literals before interpreting their contents as string delimiters
  `.github/workflows/pr-triage.yml:224`

  The lexical pre-pass newly misses real tests when an earlier added line contains a double-quote character literal. For example, add this entire block to a `crates/**/src/*.rs` file outside a test-named path, with no other recognized test attribute in the PR:

  ```rust
  const QUOTE: char = '"';
  #[test]
  fn real_test() {
      assert_eq!(QUOTE as u32, 34);
  }
  ```

  This compiles under `rustc --test` and registers and passes one test. The detector at the reviewed merge-base/current target (`bfc44f9`) returns `true`, while this PR head (`d800f08`) returns `false`. The same regression occurs with a byte-character literal such as `b'"'`. All relevant lines are consecutive additions, so the reproduction is inside the scanner's stated tracking boundary.

  **Root cause and effect:** the pre-pass reaches the `"` inside the character token and sets `inString = true`. It never recognizes the surrounding apostrophes as a character-literal boundary, so that state survives onto the following `#[test]` line. The line enters `nonCodeStart`, and the candidate scan skips it before checking the attribute. For an author subject to triage, with no other test signal, `touchedTests` remains false: the workflow applies or retains `needs-tests` and may post the one-time “no tests changed” guidance. This is an advisory-label correctness regression, not a failure of the Rust test itself.

  **Requested correction:** address the literal-boundary handling in the pre-pass so contents of a valid character or byte-character token cannot open string state. One bounded approach is to recognize a complete character token, including its escape form, before interpreting its contents as string/comment syntax. The implementation is your choice; the required outcome is that the example and equivalent byte/escaped-quote forms retain their previous detection result. Avoid a general “skip from apostrophe to next apostrophe” rule: Rust lifetimes and labels also use apostrophes, and that could hide subsequent source. Keep the existing string/comment filtering effective.

  **Focused validation:** add cases to `scripts/test-pr-triage-detect.mjs` with character and byte-character double quotes, including escaped forms, followed by a real test in the same added run; these should return `true` and expose the current regression. Include a lifetime or label compatibility case for any apostrophe-handling change, and retain the existing cases that reject test-looking text inside strings/comments. Run `node scripts/test-pr-triage-detect.mjs`, including its existing adversarial probes. The current 38-case matrix passes but does not cover this character-token boundary.

  Keep this repair within the existing consecutive-added-line model and bounded scan. It does not require tracking state across unchanged lines, resolving unknown pre-hunk context, expanding the supported harness names, or replacing the detector with a full Rust parser.

**Comment by beardthelion** [comment]
Recognize character literals before interpreting their contents as string delimiters
`.github/workflows/pr-triage.yml:224`

Addressed. The pre-pass now recognizes Rust char and byte-char literals (`'X'`, `'\X'`, `'\u{...}'`, `b'X'`) before interpreting their contents as string or comment syntax. On seeing `'`, it does a bounded lookahead for a closing `'`: 2-4 chars for simple and escaped forms, or to `}` for `\u` escapes (bounded to 20 chars, at most 6 hex digits). If found, the whole literal is skipped so a `"` inside it cannot open string state. If not found, the token is a lifetime or label and the apostrophe is skipped as a regular char. This is not a skip-to-next-`'` rule; lifetimes and labels fall through unchanged.

Added 8 test cases: char and byte-char with double quotes, escaped quote, escaped backslash, unicode escape, lifetime, label, and static lifetime, each followed by a real `#[test]` in the same added run. All return `true`. Existing 38 cases retained. `node scripts/test-pr-triage-detect.mjs` reports 47 cases passed.

Also bounded the `\u{` escape scan to 20 chars past `{`. Without the bound, a patch repeating `'\u{` thousands of times made the scan quadratic (21s for 30000 reps); with the bound it is 22ms. Added a runtime probe for that shape.

**Review by jatmn — CHANGES_REQUESTED** [review]
wrong details.

**Review by jatmn — APPROVED** [review]
lgtm
